### DISCUSSION TITLE
S3: put_object doesn't yield to tokio runtime

### DISCUSSION THREAD

**Discussion by rtdavis22** [awslabs/aws-sdk-rust#604] [General]
Hi, I'm calling `put_object` with a `ByteStream` that reads from a file. The file is very large so it takes several seconds for the upload to complete. It appears that during the upload, execution is never yielded back to the Tokio runtime. Is there any way that I can yield to the Tokio runtime as the file is uploading?

**Comment by jdisanti** [comment]
That's odd. Can you provide more details about your Tokio runtime? Are you using the single threaded runtime? What symptoms are you observing that are leading you to conclude that the upload is blocking?

**Reply by rtdavis22** [reply]
@jdisanti My Tokio runtime is multi-threaded with several worker threads. When we do the upload, the runtime seems to be blocked since the server is no longer able to service requests until the upload is complete. I think what we're noticing is the issue described here: https://github.com/tokio-rs/tokio/issues/4730.

I took a look at what happens in the upload code and it doesn't appear that there are `.await`s while the file is being uploaded, so I think the upload is blocking the Tokio runtime.

To resolve the issue, we used `.send()` in a blocking thread:
```
tokio::task::spawn_blocking(|| futures::executor::block_on(request.send()));
```
But it seems like there should be `.await`s while the file is being uploaded so we wouldn't need to do this.

**Reply by konnextv** [reply]
@rtdavis22 Do you by chance know of any changes in aws-sdk-s3 that mitigate this issue? I think I just encountered this by chance and can't really believe that this has been a known issue for 2 years.

**Reply by Velfi** [reply]
I wouldn't call this a "known issue" because there is no minimal reproducible example and it's not clear if this is an issue with `tokio`, the SDK, or the user's code. If you're having an issue, we'd be happy to help so long as we can reproduce it.

**Reply by konnextv** [reply]
Thanks for your feedback @Velfi. From reading the tokio issue that is linked above I can imagine this will be a hard one to reproduce.
I will dig into this tomorrow, but the outline of what I am doing in a loop (throttled by `tokio::time::sleep` and with a oneshot channel that is monitored using the `tokio::select!` macro) is the following:
- iterate over matching files in a directory, then for each file:
- do `ByteStream::from_path(&path).await`
- create a `PubObjectFluentBuilder` supplying the bytestream (with `.unwrap()`)
- send the request with `request.send().await`

Here is why I think what I witnessed matches what the tokio issue is about. The loop was not exited when the oneshot channel was fired which can be explained by blocking.
NOTE: I am kind of new to Rust. While writing this I realize that I am using potentially blocking filesystem functions inside the spawned future - from what I learned so far this is not a good practice but the filesystem should not block _indefinitely_ (correct me if I am wrong). If I am wrong on this, this is my error - I do not want to turn this discussion into a lesson on blocking functions but maybe this will help someone running into this in the future.

**Reply by Velfi** [reply]
I'd be curious to know
- How many files are you dealing with?
- What is the median size of files + are there any outliers?
- How are you limiting concurrency?
- Where is this code running? What are the specs of the runner.

**Reply by konnextv** [reply]
- _How many files are you dealing with?_
1-2 per iteration (apart from initial exports when there could be a few hundred files)
- _What is the median size of files + are there any outliers?_
~1.5MB per file with few exceptions where files are smaller
- _How are you limiting concurrency?_
There is one `async fn` that iterates over the files. That is run in a `loop` with a sleep/cancel after. I was using `rt-multi-thread` but 
switched to `rt` after realizing that I do not need multithreading.
- _Where is this code running? What are the specs of the runner?_
An on-premise Windows 10 workstation. CPUs are Intel i5/i7 with around 12 threads.
